### ISSUE TITLE
macro used in if condition triggers block_in_if_condition_stmt

### ISSUE THREAD

**Issue by karyon** [Manishearth/rust-clippy#528]
in servo, this line
 ```
if log_enabled!(log::LogLevel::Info) {
```
gives me this warning:
```
/servo/components/style/selector_matching.rs:39:13: 39:46 warning: in an 'if' condition, avoid complex blocks or closures with blocks; instead, move the block or closure higher and bind it with a 'let', #[warn(block_in_if_condition_stmt)] on by default
/servo/components/style/selector_matching.rs:39          if log_enabled!(log::LogLevel::Info) {
                                                            ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/servo/components/style/selector_matching.rs:39:13: 39:46 note: in this expansion of log_enabled! (defined in <log macros>)
/servo/components/style/selector_matching.rs:39:13: 39:46 help: try
let res = {
let lvl = $ lvl ; lvl <= $ crate:: __static_max_level (  ) && lvl <= $ crate::
max_log_level (  ) && $ crate:: __enabled ( lvl , $ target ) };
if res {
    let location = input.source_location(position);
    info!("{}:{} {}", location.line, location.column, message)
} ... 
for further information visit https://github.com/Manishearth/rust-clippy/wiki#block_in_if_condition_stmt
```
which i think is not intended.

**Comment by llogiq** [comment]
That lint needs a macro check.

**Comment by karyon** [comment]
same warning appears on this code:
```
        assert!(LIVE_REFERENCES.with(|ref r| {
            let r = r.borrow();
            let live_references = r.as_ref().unwrap();
            self.owner_thread == (&*live_references) as *const _ as *const libc::c_void
        }));
```
if even says the if is inside the macro expansion.

**Comment by karyon** [comment]
after cherry-picking your fix, i still get this warning, although by now i'm really not sure whether it's another problem of my servo branch...

**Comment by karyon** [comment]
now i'm pretty sure i'm on 4484448cd1b879b361d0633d8a4188f46519e8c1 and still get this false positive. for example in `components/style/selector_matching.rs:39`
```
    if log_enabled!(log::LogLevel::Info) {
```
and `components/script/dom/bindings/refcounted.rs:94`
```
        assert!(LIVE_REFERENCES.with(|ref r| {
            let r = r.borrow();
            let live_references = r.as_ref().unwrap();
            self.owner_thread == (&*live_references) as *const _ as *const libc::c_void
        }));
```

**Comment by Manishearth** [comment]
Perhaps we should just add a strong macro check?

**Comment by karyon** [comment]
@Manishearth ping, is this supposed to stay closed?

**Comment by Manishearth** [comment]
I already fixed it in https://github.com/Manishearth/rust-clippy/commit/a21108a2963c63310d9c5b452e7faf3e622c47de

**Comment by karyon** [comment]
crazy, thanks.

**Comment by yberreby** [comment]
@Manishearth Strangely enough, I'm having this issue even though you fixed it. I'm using latest clippy (0.0.37). Tried both the regular way (`#![plugin(clippy)]`) and via `cargo clippy`, no luck.

```
foo/client.rs:60:12: 60:41 warning: in an 'if' condition, avoid complex blocks or closures with blocks; instead, move the block or closure higher and bind it with a 'let', #[warn(block_in_if_condition_stmt)] on by default
foo/client.rs:60         if log_enabled!(LogLevel::Debug) {
                                 ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~
foo/client.rs:60:12: 60:41 note: in this expansion of log_enabled! (defined in <log macros>)
foo/client.rs:60:12: 60:41 help: try
let res = {
let lvl = $ lvl ; lvl <= $ crate:: __static_max_level (  ) && lvl <= $ crate::
max_log_level (  ) && $ crate:: __enabled ( lvl , $ target ) };
if res {
    let mut response_body = String::new();
    try!(res.read_to_string(&mut response_body));
    debug!("foo:\n{}", response_body);
} ...
```

**Comment by llogiq** [comment]
I think the problem lies within the macro check. I'm currently testing a variation that only checks if the block is expanded from a macro.

**Comment by llogiq** [comment]
I'll see if adding a macro check to the visitor can fix this.

**Comment by llogiq** [comment]
I was wrong; the problem is not within the Visitor, but in the macro check in line 96. It fails to find the macro expansion.
